pkg/volume/scaleio/sio_util.go

```diff
@@ -0,0 +1,249 @@
+		configMap = &api.ConfigMap{
+			ObjectMeta: meta.ObjectMeta{
+				Name:      configName,
+				Namespace: namespace,
+			},
+			Data: data,
+		}
+		configMap, err = kubeClient.Core().ConfigMaps(namespace).Create(configMap)
+		glog.V(4).Info(log("creating new configMap  %s/%s", namespace, configName))
+	} else {
+		configMap.Data = data
+		configMap, err = kubeClient.Core().ConfigMaps(namespace).Update(configMap)
+		glog.V(4).Info(log("updating existing configMap %s/%s", namespace, configName))
+	}
+	if err != nil {
+		glog.Error(log("failed to save configMap: %v", err))
+		return err
+	}
+	glog.V(4).Info(log("config map %s saved successfully", configMap.Name))
+	return nil
+}
+
+// setupConfigData setups the needed config data as internal configMap.
+func setupConfigData(plug *sioPlugin, namespace string, configData map[string]string) error {
+	glog.V(4).Info(log("setting up config data"))
+	// ensure needed config is provided
+	if err := validateConfigs(configData); err != nil {
+		glog.Error(log("config data setup failed: %s", err))
+		return err
+	}
+
+	// load secret
+	secretRefName := configData[confKey.secretRef]
+	glog.V(4).Info(log("loading secret %s in config data", secretRefName))
+	kubeClient := plug.host.GetKubeClient()
+	secretMap, err := volutil.GetSecretForPV(namespace, secretRefName, sioPluginName, kubeClient)
+	if err != nil {
+		glog.Error(log("failed to get secret: %v", err))
+		return secretNotFoundErr
+	}
+	// merge secret data
+	for key, val := range secretMap {
+		configData[key] = val
+	}
+
+	return nil
+}
+
+// loadConfigData retrieves configuration data from storage
+func loadConfigMap(plug *sioPlugin, namespace, configName string) (map[string]string, error) {
+	glog.V(4).Info(log("loading configMap %s/%s", namespace, configName))
+	kubeClient := plug.host.GetKubeClient()
+	confMap, err := kubeClient.Core().ConfigMaps(namespace).Get(configName, meta.GetOptions{})
+	if err != nil {
+		glog.Error(log("failed to load ConfigMap %s: %v", configName, err))
+		return nil, configMapNotFoundErr
+	}
+
+	configs := confMap.Data
+	applyConfigDefaults(configs)
+	if err := validateConfigs(configs); err != nil {
+		glog.Error(log("failed to load ConfigMap %s: %v", err))
+		return nil, err
+	}
+
+	return configs, nil
+}
+
+// getVolumeSourceFromSpec safely extracts ScaleIOVolumeSource from spec
+func getVolumeSourceFromSpec(spec *volume.Spec) (*api.ScaleIOVolumeSource, error) {
+	if spec.Volume != nil && spec.Volume.ScaleIO != nil {
+		return spec.Volume.ScaleIO, nil
+	}
+	if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.ScaleIO != nil {
+		return spec.PersistentVolume.Spec.ScaleIO, nil
+	}
+
+	return nil, fmt.Errorf("ScaleIO not defined in spec")
+}
+
+// splitVolumeName extracts the namespace and volumeName from vol
+// where vol = <namespace>nsSep<volumeName>
+func splitVolumeName(vol string) (namespace, volName string) {
```

Comment:
@rootfs yes. In the rare cases when `UnMounter` is created after a server restart and the controller starts to garbage-collect orphaned pods, the code checks to see if it can extract the namespace from the volume.